### ISSUE TITLE
ca.cer fails to download over https

### ISSUE THREAD

**Issue by leomoty** [cockpit-project/cockpit#13174]
Cockpit version: 206
OS: Fedora Server Edition

Yesterday I had some misconfiguration and got greeted with the certificate page to import it locally. However since it was loading the page over `https`, at least chrome wouldn't download the file at all:

![image](https://user-images.githubusercontent.com/1463740/69260582-e3a1ee00-0b9e-11ea-9abd-6ae93e16e34d.png)

Failed - Invalid certificate

Maybe downgrade it to http if that page is hit?

**Comment by marusak** [comment]
@mvollmer Is this the new feature that you worked on a few months ago? If so, do you have any knowledge that https could cause problem with downloading?

**Comment by garrett** [comment]
## Quick summary

This page probably doesn't apply to your scenario. If you got this far in a normal browser (that is: not on iOS), then the same certificate applies to both normal HTTPS connections and websockets.

To see this page, you'd need to have a certificate that works enough for HTTPS but not websockets (which is mainly just an iOS bug).

(Or perhaps @martinpitt and @mvollmer might know other corner cases where it could show up?)


## Explanation

The purpose of downloading the certificate over HTTPS is mainly to support browsers that support self-signed certificates for HTTPS but not for websockets unless the certificate is installed manually.

In other words: Safari on iOS doesn't support self-signed certficiates for websockets, and this is a workaround for iOS. (Sadly this works in the Simulator, but I couldn't get it working on an actual iPhone.)

As you're using Chrome, this probably doesn't apply to you, as Chrome uses the same certificates for HTTPS and websockets. (Except on iOS, because "Chrome" on iOS is actually just WebKit, just like every other browser on iOS is basically just Safari with a different UI wrapped around it.)

The page where you see the "ca.cer" download link is confusing, but it usually isn't seen under most circumstances. It should probably be reworded nonetheless.

**Comment by garrett** [comment]
(Even shorter answer: If you manually import the certificate that is linked, it would probably have the same effect as accepting it in-browser under most circumstances.)

**Comment by martinpitt** [comment]
@leomoty : The only way I can currently imagine that would get you there is if you use cockpit behind a reverse proxy, and didn't configure it to allow an upgrade to websocket (see https://github.com/cockpit-project/cockpit/wiki/Proxying-Cockpit-over-NGINX). Is that what happened?

**Comment by leomoty** [comment]
@garrett I am aware this is not useful for my case.

@martinpitt not really, this is a Virtual Machine running at my work machine. And I am forced to use Kaspersky which usually self signs a certificate for the domain. However yesterday it didn't.

And since the `ca.cer` link appeared and I remembered the discussion, I decided to try to download it. The thing is... it tried to load `https://ip:9090/ca.cer` which fails in chrome because the certificate is not valid and it gives you no option to deal with. Trying to load `http://ip:9090/ca.cer` works instead.

**Comment by martinpitt** [comment]
Indeed the notion of downloading https://ip:9090/ca.cer makes no sense -- if you could  do that, then you already trust the tls certificate in your browser, and you wouldn't need to install it any more. This is indeed meant for downloading through http:// (what can you do, pull yourself out of the swamp :cry: )
